amqp/src/main/scala/akka/stream/alpakka/amqp/impl/AmqpAsyncFlow.scala

```diff
@@ -0,0 +1,338 @@
+    }
+  }
+
+  private def onConfirmation(tag: DeliveryTag, multiple: Boolean): Unit = {
+    log.debug("Received confirmation for deliveryTag {} (multiple={}).", tag, multiple)
+
+    val dequeued: Iterable[AwaitingMessage[T]] = dequeueAwaitingMessages(tag, multiple)
+
+    dequeued.foreach(m => cancelTimer(m.tag))
+
+    pushOrEnqueueResults(
+      dequeued.map(m => WriteResult.confirmed(m.passThrough))
+    )
+  }
+
+  private def onRejection(tag: DeliveryTag, multiple: Boolean): Unit = {
+    log.debug("Received rejection for deliveryTag {} (multiple={}).", tag, multiple)
+
+    val dequeued: Iterable[AwaitingMessage[T]] = dequeueAwaitingMessages(tag, multiple)
+
+    dequeued.foreach(m => cancelTimer(m.tag))
+
+    pushOrEnqueueResults(
+      dequeued.map(m => WriteResult.rejected(m.passThrough))
+    )
+  }
+
+  private def pushOrEnqueueResults(results: Iterable[WriteResult[T]]): Unit = {
+    results.foreach(result => {
+      if (exitQueue.nonEmpty || !isAvailable(out)) {
+        log.debug("Message {} queued for downstream push.", result)
+        exitQueue.enqueue(result)
+      } else {
+        log.debug("Pushing {} downstream.", result)
+        push(out, result)
+      }
+    })
+    if (isFinished) closeStage()
+  }
+
+  override def postStop(): Unit = {
+    promise.tryFailure(new RuntimeException("Stage stopped unexpectedly."))
+    super.postStop()
+  }
+
+  override def onFailure(ex: Throwable): Unit = {
+    promise.tryFailure(ex)
+    onFailure(ex)
+  }
+
+  def dequeueAwaitingMessages(tag: DeliveryTag, multiple: Boolean): Iterable[AwaitingMessage[T]]
+
+  def enqueueMessage(tag: DeliveryTag, passThrough: T): Unit
+
+  def messagesAwaitingDelivery: Int
+
+  def noAwaitingMessages: Boolean
+
+  setHandler(
+    in,
+    new InHandler {
+
+      override def onPush(): Unit = {
+        val message: WriteMessage[T] = grab(in)
+        val tag = publish(message)
+
+        scheduleOnce(tag, confirmationTimeout)
+        enqueueMessage(tag, message.passThrough)
+        if (messagesAwaitingDelivery + exitQueue.size < bufferSize && !hasBeenPulled(in)) tryPull(in)
+      }
+
+      override def onUpstreamFailure(ex: Throwable): Unit = {
+        upstreamException = Some(ex)
+        if (isFinished)
+          closeStage()
+        else
+          log.debug("Received upstream failure signal - stage will be failed when all buffered messages are processed")
+
+      }
+
+      override def onUpstreamFinish(): Unit =
+        if (noAwaitingMessages && exitQueue.isEmpty) {
+          promise.success(Done)
+          super.onUpstreamFinish()
+        } else {
+          log.debug("Received upstream finish signal - stage will be closed when all buffered messages are processed")
+        }
+
+      private def publish(message: WriteMessage[T]): DeliveryTag = {
+        val tag: DeliveryTag = channel.getNextPublishSeqNo
+
+        log.debug("Publishing message {} with deliveryTag {}.", message, tag)
+
+        Try(
```

Comment:
what's the reason using `Try` class and extract it right here, why not just `try` clause?

Reply:
Well, you are right, I'm not sure why I have done it this way (it was some time ago), but anyway it looks like throwing exception from `onPush` has the same effect as failing the stage manually, so there is no need to handle this exception here at all.